File: time.py
# import sqlite3
# import time
# import matplotlib.pyplot as plt
# import numpy as np
# import random
# from datetime import datetime, timedelta

# # Настройки эксперимента
# max_size = 6  # Максимальный размер таблицы (10^6 записей)
# sizes = []  # Размеры таблиц
# num_samples = 50  # Количество замеров для усреднения

# # Списки для хранения результатов
# times_without_index = []
# times_with_index = []

# # Минимальное время для отображения на графике
# MIN_TIME = 1e-9  # 1 наносекунда

# # Генерация размеров таблиц
# for size in [10**i for i in range(1, max_size + 1)]:
#     sizes.append(size)
#     # Добавляем дополнительные размеры между степенями
#     if size != 10**max_size:
#         for additional_size in range(size, size*10, size):
#             sizes.append(additional_size)

# sizes = sorted(set(sizes))  # Удаляем дубликаты и сортируем
# print(sizes, len(sizes))

# def generate_random_timestamp(start, end):
#     """Генерация случайной временной метки между start и end"""
#     delta = end - start
#     int_delta = (delta.days * 24 * 60 * 60) + delta.seconds
#     random_second = random.randrange(int_delta)
#     return start + timedelta(seconds=random_second)

# for size in sizes:
#     conn = sqlite3.connect(':memory:')
#     cursor = conn.cursor()
    
#     # Создание таблиц (упрощенная версия вашей схемы)
#     cursor.execute('''
#         CREATE TABLE movies (
#             id TEXT PRIMARY KEY,
#             title TEXT NOT NULL,
#             duration TEXT NOT NULL
#         )
#     ''')
    
#     cursor.execute('''
#         CREATE TABLE halls (
#             id TEXT PRIMARY KEY,
#             name TEXT NOT NULL
#         )
#     ''')
    
#     cursor.execute('''
#         CREATE TABLE movie_shows (
#             id TEXT PRIMARY KEY,
#             movie_id TEXT REFERENCES movies(id),
#             hall_id TEXT REFERENCES halls(id),
#             start_time TEXT NOT NULL,
#             language TEXT NOT NULL
#         )
#     ''')
    
#     # Заполнение таблиц тестовыми данными
#     # Сначала создаем фильмы и залы
#     movies = [('movie_' + str(i), f'Movie {i}', '02:30:00') for i in range(100)]
#     halls = [('hall_' + str(i), f'Hall {i}') for i in range(10)]
    
#     cursor.executemany('INSERT INTO movies VALUES (?, ?, ?)', movies)
#     cursor.executemany('INSERT INTO halls VALUES (?, ?)', halls)
    
#     # Генерируем сеансы
#     start_date = datetime(2023, 1, 1)
#     end_date = datetime(2023, 12, 31)
    
#     movie_shows = []
#     for i in range(size):
#         movie_id = random.choice(movies)[0]
#         hall_id = random.choice(halls)[0]
#         start_time = generate_random_timestamp(start_date, end_date)
#         language = random.choice(['English', 'Spanish', 'French', 'German', 'Italian', 'Русский'])
#         movie_shows.append((f'show_{i}', movie_id, hall_id, start_time.isoformat(), language))
    
#     cursor.executemany('INSERT INTO movie_shows VALUES (?, ?, ?, ?, ?)', movie_shows)
#     conn.commit()
    
#     # Тестирование запроса по hall_id и start_time БЕЗ индекса
#     times = []
#     for _ in range(num_samples):
#         hall_id = random.choice(halls)[0]
#         random_time = generate_random_timestamp(start_date, end_date)
        
#         start = time.perf_counter()
#         cursor.execute('''
#             SELECT * FROM movie_shows 
#             WHERE hall_id = ? AND start_time >= ? AND start_time <= ?
#         ''', (hall_id, 
#               (random_time - timedelta(hours=2)).isoformat(), 
#               (random_time + timedelta(hours=2)).isoformat()))
#         cursor.fetchall()
#         elapsed = max(time.perf_counter() - start, MIN_TIME)
#         times.append(elapsed)
#     avg_time = np.mean(times)
#     times_without_index.append(avg_time)
    
#     # Создание индекса
#     cursor.execute('CREATE INDEX idx_movie_shows_hall_time ON movie_shows(hall_id, start_time)')
    
#     # Тестирование запроса по hall_id и start_time С индексом
#     times = []
#     for _ in range(num_samples):
#         hall_id = random.choice(halls)[0]
#         random_time = generate_random_timestamp(start_date, end_date)
        
#         start = time.perf_counter()
#         cursor.execute('''
#             SELECT * FROM movie_shows 
#             WHERE hall_id = ? AND start_time >= ? AND start_time <= ?
#         ''', (hall_id, 
#               (random_time - timedelta(hours=2)).isoformat(), 
#               (random_time + timedelta(hours=2)).isoformat()))
#         cursor.fetchall()
#         elapsed = max(time.perf_counter() - start, MIN_TIME)
#         times.append(elapsed)
#     avg_time = np.mean(times)
#     times_with_index.append(avg_time)
    
#     conn.close()

# # Построение графика
# plt.figure(figsize=(12, 7))
# plt.plot(sizes, times_without_index, color='red', linestyle='-', label='Без индекса')
# plt.plot(sizes, times_with_index, color='green', linestyle='--', label='С индексом')
# plt.xscale('log')
# plt.yscale('log')  # Логарифмическая шкала для оси Y
# plt.xlabel('Размер таблицы (количество сеансов)')
# plt.ylabel('Среднее время выполнения запроса (сек)')
# plt.title('Производительность запросов к таблице movie_shows\n(поиск по hall_id и временному диапазону)')
# plt.legend()
# plt.grid(True, which="both", ls="--")
# plt.show()

import psycopg2
import time
import matplotlib.pyplot as plt
import numpy as np
import random
from datetime import datetime, timedelta
from psycopg2 import sql

# Настройки эксперимента
max_size = 5  # Максимальный размер таблицы (10^6 записей)
sizes = []  # Размеры таблиц
num_samples = 5  # Количество замеров для усреднения

# Списки для хранения результатов
times_without_index = []
times_with_index = []

# Минимальное время для отображения на графике
MIN_TIME = 1e-9  # 1 наносекунда

# Генерация размеров таблиц
for size in [10**i for i in range(1, max_size + 1)]:
    sizes.append(size)
    # Добавляем дополнительные размеры между степенями
    if size != 10**max_size:
        for additional_size in range(size, size*10, size):
            sizes.append(additional_size)

sizes = sorted(set(sizes))  # Удаляем дубликаты и сортируем
print(sizes, len(sizes))

def generate_random_timestamp(start, end):
    """Генерация случайной временной метки между start и end"""
    delta = end - start
    int_delta = (delta.days * 24 * 60 * 60) + delta.seconds
    random_second = random.randrange(int_delta)
    return start + timedelta(seconds=random_second)

# Подключение к PostgreSQL
conn = psycopg2.connect(
    dbname="cinema",
    user="postgres",
    password="postgres",
    host="localhost"
)
conn.autocommit = True
cursor = conn.cursor()

for size in sizes:
    print(size)
    try:
        # Создание временной схемы для изоляции экспериментов
        schema_name = f"experiment_{size}"
        cursor.execute(sql.SQL("CREATE SCHEMA IF NOT EXISTS {}").format(sql.Identifier(schema_name)))
        
        # Создание таблиц
        cursor.execute(sql.SQL('''
            CREATE TABLE IF NOT EXISTS {}.movies (
                id TEXT PRIMARY KEY,
                title TEXT NOT NULL,
                duration TEXT NOT NULL
            )
        ''').format(sql.Identifier(schema_name)))
        
        cursor.execute(sql.SQL('''
            CREATE TABLE IF NOT EXISTS {}.halls (
                id TEXT PRIMARY KEY,
                name TEXT NOT NULL
            )
        ''').format(sql.Identifier(schema_name)))
        
        cursor.execute(sql.SQL('''
            CREATE TABLE IF NOT EXISTS {}.movie_shows (
                id TEXT PRIMARY KEY,
                movie_id TEXT REFERENCES {}.movies(id),
                hall_id TEXT REFERENCES {}.halls(id),
                start_time TIMESTAMP NOT NULL,
                language TEXT NOT NULL
            )
        ''').format(
            sql.Identifier(schema_name),
            sql.Identifier(schema_name),
            sql.Identifier(schema_name)
        ))
        
        # Заполнение таблиц тестовыми данными
        # Сначала создаем фильмы и залы
        movies = [('movie_' + str(i), f'Movie {i}', '02:30:00') for i in range(100)]
        halls = [('hall_' + str(i), f'Hall {i}') for i in range(10)]
        
        cursor.executemany(
            sql.SQL('INSERT INTO {}.movies VALUES (%s, %s, %s)').format(sql.Identifier(schema_name)),
            movies
        )
        cursor.executemany(
            sql.SQL('INSERT INTO {}.halls VALUES (%s, %s)').format(sql.Identifier(schema_name)),
            halls
        )
        
        # Генерируем сеансы
        start_date = datetime(2023, 1, 1)
        end_date = datetime(2023, 12, 31)
        
        # Используем batch-вставку для ускорения
        batch_size = 10000
        for i in range(0, size, batch_size):
            batch = []
            for j in range(i, min(i + batch_size, size)):
                movie_id = random.choice(movies)[0]
                hall_id = random.choice(halls)[0]
                start_time = generate_random_timestamp(start_date, end_date)
                language = random.choice(['English', 'Spanish', 'French', 'German', 'Italian', 'Русский'])
                batch.append((f'show_{j}', movie_id, hall_id, start_time, language))
            
            cursor.executemany(
                sql.SQL('INSERT INTO {}.movie_shows VALUES (%s, %s, %s, %s, %s)').format(sql.Identifier(schema_name)),
                batch
            )
        
        # Анализируем таблицы для обновления статистики
        cursor.execute(sql.SQL('ANALYZE {}.movies').format(sql.Identifier(schema_name)))
        cursor.execute(sql.SQL('ANALYZE {}.halls').format(sql.Identifier(schema_name)))
        cursor.execute(sql.SQL('ANALYZE {}.movie_shows').format(sql.Identifier(schema_name)))
        
        # Тестирование запроса по hall_id и start_time БЕЗ индекса
        times = []
        for _ in range(num_samples):
            hall_id = random.choice(halls)[0]
            random_time = generate_random_timestamp(start_date, end_date)
            
            start = time.perf_counter()
            cursor.execute(
                sql.SQL('''
                    SELECT * FROM {}.movie_shows 
                    WHERE hall_id = %s AND start_time >= %s AND start_time <= %s
                ''').format(sql.Identifier(schema_name)),
                (hall_id, random_time - timedelta(hours=2), random_time + timedelta(hours=2))
            )
            cursor.fetchall()
            elapsed = max(time.perf_counter() - start, MIN_TIME)
            times.append(elapsed)
        avg_time = np.mean(times)
        times_without_index.append(avg_time)
        
        # Создание индекса
        cursor.execute(
            sql.SQL('CREATE INDEX IF NOT EXISTS idx_movie_shows_hall_time ON {}.movie_shows(hall_id, start_time)')
            .format(sql.Identifier(schema_name))
        )
        
        # Тестирование запроса по hall_id и start_time С индексом
        times = []
        for _ in range(num_samples):
            hall_id = random.choice(halls)[0]
            random_time = generate_random_timestamp(start_date, end_date)
            
            start = time.perf_counter()
            cursor.execute(
                sql.SQL('''
                    SELECT * FROM {}.movie_shows 
                    WHERE hall_id = %s AND start_time >= %s AND start_time <= %s
                ''').format(sql.Identifier(schema_name)),
                (hall_id, random_time - timedelta(hours=2), random_time + timedelta(hours=2))
            )
            cursor.fetchall()
            elapsed = max(time.perf_counter() - start, MIN_TIME)
            times.append(elapsed)
        avg_time = np.mean(times)
        times_with_index.append(avg_time)
        
    finally:
        # Удаление временной схемы
        cursor.execute(sql.SQL('DROP SCHEMA IF EXISTS {} CASCADE').format(sql.Identifier(schema_name)))

cursor.close()
conn.close()

# Построение графика
plt.figure(figsize=(12, 7))
plt.plot(sizes, times_without_index, color='red', linestyle='-', label='Без индекса')
plt.plot(sizes, times_with_index, color='green', linestyle='--', label='С индексом')
plt.xscale('log')
plt.yscale('log')  # Логарифмическая шкала для оси Y
plt.xlabel('Размер таблицы (количество сеансов)')
plt.ylabel('Среднее время выполнения запроса (сек)')
plt.title('Производительность запросов к таблице movie_shows\n(поиск по hall_id и временному диапазону)')
plt.legend()
plt.grid(True, which="both", ls="--")
plt.show()

File: time.go
package main

import (
	"database/sql"
	"fmt"
	"log"
	"math/rand"
	"time"

	"gonum.org/v1/plot"
	"gonum.org/v1/plot/plotter"
	"gonum.org/v1/plot/plotutil"
	"gonum.org/v1/plot/vg"

	_ "github.com/jackc/pgx/v5/stdlib"
)

const (
	maxSize    = 6                   // Максимальный размер таблицы (10^6 записей)
	numSamples = 50                  // Количество замеров для усреднения
	minTime    = 1 * time.Nanosecond // Минимальное время для отображения
	dbConnStr  = "host=localhost port=5432 user=postgres password=postgres dbname=cinema sslmode=disable"
)

func time_all() {
	// Генерация размеров таблиц
	sizes := generateSizes(maxSize)

	// Результаты
	timesWithoutIndex := make([]float64, len(sizes))
	timesWithIndex := make([]float64, len(sizes))

	// Подключение к БД
	db, err := sql.Open("pgx", dbConnStr)
	if err != nil {
		log.Fatalf("Unable to connect to database: %v\n", err)
	}
	defer db.Close()

	println(len(sizes), sizes)
	// Запуск тестов
	for i, size := range sizes {
		println("start", i, size)
		log.Printf("Testing with size: %d\n", size)

		// Подготовка тестовых данных
		if err := prepareTestData(db, size); err != nil {
			log.Fatal(err)
		}

		// Тест без индекса
		avgTime, err := testQueryPerformance(db, size, false)
		if err != nil {
			log.Fatal(err)
		}
		timesWithoutIndex[i] = avgTime.Seconds()

		// Создание индекса
		if _, err := db.Exec("CREATE INDEX IF NOT EXISTS idx_movie_shows_hall_time ON movie_shows(hall_id, start_time)"); err != nil {
			log.Fatal(err)
		}

		// Тест с индексом
		avgTime, err = testQueryPerformance(db, size, true)
		if err != nil {
			log.Fatal(err)
		}
		timesWithIndex[i] = avgTime.Seconds()

		// Очистка данных
		if _, err := db.Exec("TRUNCATE TABLE movie_shows, movies, halls CASCADE"); err != nil {
			log.Fatal(err)
		}

		println("end", i, size)
	}

	// Построение графика
	if err := plotResults(sizes, timesWithoutIndex, timesWithIndex); err != nil {
		log.Fatal(err)
	}
}

func generateSizes(maxSize int) []int {
	var sizes []int
	for i := 1; i <= maxSize; i++ {
		power := 1
		for j := 0; j < i; j++ {
			power *= 10
		}
		sizes = append(sizes, power)

		// Добавляем промежуточные значения
		if i < maxSize {
			for j := 2; j <= 9; j++ {
				sizes = append(sizes, power*j)
			}
		}
	}
	return sizes
}

func prepareTestData(db *sql.DB, size int) error {
	// Создание тестовых фильмов и залов
	if _, err := db.Exec(`
        INSERT INTO movies (id, title, duration, description, age_limit, release_date)
        SELECT 
            gen_random_uuid(),
            'Movie ' || seq,
            '02:30:00'::time,
            'Description for movie ' || seq,
            12,
            NOW() - (random() * 365 * 5 || ' days')::interval
        FROM generate_series(1, 100) seq
        ON CONFLICT DO NOTHING
    `); err != nil {
		return fmt.Errorf("inserting movies: %w", err)
	}

	// Создание screen_types
	if _, err := db.Exec(`
        INSERT INTO screen_types (id, name, description)
        VALUES 
            (gen_random_uuid(), 'Standard', 'Standard screen'),
            (gen_random_uuid(), 'IMAX', 'IMAX screen')
        ON CONFLICT DO NOTHING
    `); err != nil {
		return fmt.Errorf("inserting screen types: %w", err)
	}

	// Создание залов
	if _, err := db.Exec(`
        INSERT INTO halls (id, name, screen_type_id, description)
        SELECT 
            gen_random_uuid(),
            'Hall ' || seq,
            (SELECT id FROM screen_types ORDER BY random() LIMIT 1),
            'Description for hall ' || seq
        FROM generate_series(1, 10) seq
        ON CONFLICT DO NOTHING
    `); err != nil {
		return fmt.Errorf("inserting halls: %w", err)
	}

	// Отключаем триггер для массовой вставки
	if _, err := db.Exec("ALTER TABLE movie_shows DISABLE TRIGGER check_movie_show_conflict_before_insert_or_update"); err != nil {
		return fmt.Errorf("disabling trigger: %w", err)
	}

	// Создание сеансов с фиксированным интервалом, чтобы избежать конфликтов
	if _, err := db.Exec(`
        INSERT INTO movie_shows (id, movie_id, hall_id, start_time, language)
        WITH hall_movies AS (
            SELECT 
                h.id as hall_id,
                m.id as movie_id,
                m.duration as duration,
                (ARRAY['English', 'Spanish', 'French', 'German', 'Italian', 'Русский'])[1 + (random() * 5)::int]::language_enum as language
            FROM 
                (SELECT id FROM halls LIMIT 10) h
                CROSS JOIN (SELECT id, duration FROM movies LIMIT 100) m
            ORDER BY random()
            LIMIT $1
        )
        SELECT
            gen_random_uuid(),
            hm.movie_id,
            hm.hall_id,
            NOW() + (seq * interval '4 hours'), -- Фиксированный интервал 4 часа между сеансами
            hm.language
        FROM 
            hall_movies hm,
            generate_series(0, $1-1) seq
    `, size); err != nil {
		return fmt.Errorf("inserting movie shows: %w", err)
	}

	// Включаем триггер обратно
	if _, err := db.Exec("ALTER TABLE movie_shows ENABLE TRIGGER check_movie_show_conflict_before_insert_or_update"); err != nil {
		return fmt.Errorf("enabling trigger: %w", err)
	}

	return nil
}

func testQueryPerformance(db *sql.DB, size int, withIndex bool) (time.Duration, error) {
	var totalTime time.Duration

	// Получаем список всех hall_id для тестирования
	rows, err := db.Query("SELECT id FROM halls")
	if err != nil {
		return 0, err
	}
	defer rows.Close()

	var hallIDs []string
	for rows.Next() {
		var id string
		if err := rows.Scan(&id); err != nil {
			return 0, err
		}
		hallIDs = append(hallIDs, id)
	}

	for i := 0; i < numSamples; i++ {
		// Выбираем случайный hall_id
		hallID := hallIDs[rand.Intn(len(hallIDs))]

		// Генерируем случайный временной диапазон (2 часа)
		randomHours := rand.Intn(48) - 24 // от -24 до +24 часов от текущего времени
		startTime := time.Now().Add(time.Duration(randomHours) * time.Hour)
		endTime := startTime.Add(2 * time.Hour)

		// Измеряем время выполнения запроса
		start := time.Now()

		rows, err := db.Query(`
			SELECT id, movie_id, hall_id, start_time, language 
			FROM movie_shows 
			WHERE hall_id = $1 
			AND start_time BETWEEN $2 AND $3
		`, hallID, startTime, endTime)

		elapsed := time.Since(start)
		if elapsed < minTime {
			elapsed = minTime
		}

		if err != nil {
			return 0, err
		}
		rows.Close()

		totalTime += elapsed
	}

	return totalTime / time.Duration(numSamples), nil
}

func plotResults(sizes []int, timesWithoutIndex, timesWithIndex []float64) error {
	p := plot.New()

	p.Title.Text = "Производительность запросов к таблице movie_shows"
	p.X.Label.Text = "Размер таблицы (количество сеансов)"
	p.Y.Label.Text = "Среднее время выполнения (сек)"
	p.X.Scale = plot.LogScale{}
	p.Y.Scale = plot.LogScale{}
	p.X.Tick.Marker = plot.LogTicks{}
	p.Y.Tick.Marker = plot.LogTicks{}

	// Создаем точки для графиков
	withoutIndexPoints := make(plotter.XYs, len(sizes))
	withIndexPoints := make(plotter.XYs, len(sizes))

	for i := range sizes {
		withoutIndexPoints[i].X = float64(sizes[i])
		withoutIndexPoints[i].Y = timesWithoutIndex[i]

		withIndexPoints[i].X = float64(sizes[i])
		withIndexPoints[i].Y = timesWithIndex[i]
	}

	// Добавляем линии на график
	err := plotutil.AddLinePoints(p,
		"Без индекса", withoutIndexPoints,
		"С индексом", withIndexPoints,
	)
	if err != nil {
		return err
	}

	// Сохраняем график в файл
	if err := p.Save(10*vg.Inch, 6*vg.Inch, "movie_shows_performance.png"); err != nil {
		return err
	}

	fmt.Println("График сохранен в movie_shows_performance.png")
	return nil
}
